src/dialect.ml

```diff
@@ -0,0 +1,181 @@
+    let open Dyn.Encoder in
+    record
+      [ "kind"      , Ml_kind.to_dyn kind
+      ; "extension" , string extension
+      ; "preprocess", option (pair Loc.to_dyn Action_dune_lang.to_dyn) preprocess
+      ; "format"    , option (triple Loc.to_dyn Action_dune_lang.to_dyn (list string)) format
+      ]
+end
+
+type t =
+  { name       : string
+  ; file_kinds : File_kind.t Ml_kind.Dict.t
+  }
+
+let name t = t.name
+
+let to_dyn { name ; file_kinds } =
+  let open Dyn.Encoder in
+  record
+    [ "name"      , string name
+    ; "file_kinds", Ml_kind.Dict.to_dyn File_kind.to_dyn file_kinds
+    ]
+
+let decode =
+  let open Dune_lang.Decoder in
+  let kind kind =
+    let+ loc, extension  = field "extension" (located string)
+    and+ preprocess = field_o "preprocess" (located Action_dune_lang.decode)
+    and+ format     = field_o "format" (map ~f:(fun (loc, x) -> loc, x, []) (located Action_dune_lang.decode))
+    in
+    let extension =
+      if String.contains extension '.' then
+        User_error.raise ~loc [ Pp.textf "extension string must not contain '.'" ];
+      "." ^ extension
+    in
+    { File_kind.kind ; extension ; preprocess ; format }
+  in
+  fields (let+ name = field "name" string
+          and+ impl = field "implementation" (fields (kind Ml_kind.Impl))
+          and+ intf = field "interface" (fields (kind Ml_kind.Intf))
+          in
+          { name ; file_kinds = Ml_kind.Dict.make ~intf ~impl })
+
+let extension { file_kinds = { Ml_kind.Dict.intf ; impl } ; _ } = function
+  | Ml_kind.Intf -> intf.extension
+  | Impl         -> impl.extension
+
+let preprocess { file_kinds = { Ml_kind.Dict.intf ; impl } ; _ } = function
+  | Ml_kind.Intf -> intf.preprocess
+  | Impl         -> impl.preprocess
+
+let format { file_kinds = { Ml_kind.Dict.intf ; impl } ; _ } = function
+  | Ml_kind.Intf -> intf.format
+  | Impl         -> impl.format
+
+let ocaml =
+  let format kind =
+    let flag_of_kind = function
+      | Ml_kind.Impl -> "--impl"
+      | Intf         -> "--intf"
+    in
+    let module S = String_with_vars in
+    Action_dune_lang.chdir (S.virt_var __POS__ "workspace_root")
+      (Action_dune_lang.run (S.virt __POS__ "ocamlformat")
+         [ S.virt __POS__ (flag_of_kind kind) ; S.virt_var __POS__ "input-file" ])
+  in
+  let file_kind kind extension =
+    { File_kind.
+      kind
+    ; extension
+    ; preprocess = None
+    ; format     = Some (Loc.none, format kind, [ ".ocamlformat" ; ".ocamlformat-ignore" ])
+    }
+  in
+  let intf = file_kind Ml_kind.Intf ".mli" in
+  let impl = file_kind Ml_kind.Impl ".ml"  in
+  { name       = "ocaml"
+  ; file_kinds = Ml_kind.Dict.make ~intf ~impl
+  }
+
+let reason =
+  let file_kind kind extension =
+    let module S = String_with_vars in
+    let preprocess =
+      Action_dune_lang.run (S.virt __POS__ "refmt")
+        [ S.virt     __POS__ "--print"
+        ; S.virt     __POS__ "binary"
+        ; S.virt_var __POS__ "input-file"
```

Comment:
Using `__POS__` here is wrong because in case of error the location points to `dune`'s source code. I guess we want to pass `Loc.none` somehow.

Reply:
If you don't want this, I think you can just use `make_var` and `make_text`. You can pass any location you want this way.

Reply:
Thanks, I did this change.